Parkinson/TasksAndSteps/APHParkinsonActivityViewController.m

```diff
@@ -0,0 +1,254 @@
+                                 NSLocalizedString(kMomentInDayFormatChoiceTookMyMedicine, nil),
+                                 NSLocalizedString(kMomentInDayFormatChoiceEvening, nil),
+                                 NSLocalizedString(kMomentInDayFormatChoiceNone, nil)
+                                 ];
+            ORKAnswerFormat  *format = [ORKTextChoiceAnswerFormat
+                                        choiceAnswerFormatWithStyle:ORKChoiceAnswerStyleSingleChoice
+                                                         textChoices:choices];
+            
+            ORKFormItem  *item = [[ORKFormItem alloc] initWithIdentifier:kMomentInDayFormat
+                                                                   text:NSLocalizedString(kMomentInDayFormatItemText, nil)
+                                                           answerFormat:format];
+            [stepQuestions addObject:item];
+        }
+        [step setFormItems:stepQuestions];
+        
+        NSMutableArray  *newSteps = [task.steps mutableCopy];
+        if ([newSteps count] >= 1) {
+            [newSteps insertObject:step atIndex:1];
+        }
+        replacementTask = [[ORKOrderedTask alloc] initWithIdentifier:taskTitle steps:newSteps];
+    }
+    return  replacementTask;
+}
+
+#pragma  mark  -  Step View Controller Delegate Methods
+
+- (void)stepViewControllerResultDidChange:(ORKStepViewController *)stepViewController
+{
+    NSString  *stepIdentifier = stepViewController.step.identifier;
+    if ([stepIdentifier isEqualToString:kMomentInDayStepIdentifier] == YES) {
+        ORKStepResult  *stepResult = stepViewController.result;
+        if ([stepResult.results count] > 0) {
+            id  object = [stepResult.results lastObject];
+            if ([object isKindOfClass:[ORKChoiceQuestionResult class]] == YES) {
+                ORKChoiceQuestionResult  *result = (ORKChoiceQuestionResult *)object;
+                    //
+                    //  stash the object away
+                    //
+                if (result.choiceAnswers.lastObject != nil) {
+                    NSDictionary  *dictionary = @{
+                                                  kMomentInDayChoiceAnswerKey : result.choiceAnswers.lastObject,
+                                                  kMomentInDayQuestionTypeKey : @(result.questionType),
+                                                  kMomentInDayIdentifierKey   : result.identifier,
+                                                  kMomentInDayStartDateKey    : result.startDate,
+                                                  kMomentInDayEndDateKey      : result.endDate,
+                                                  };
+                    NSUserDefaults  *defaults = [NSUserDefaults standardUserDefaults];
+                    [defaults setObject:dictionary forKey:kMomentInDayUserDefaultsKey];
+                    [defaults synchronize];
+                }
```

Comment:
> On Jun 29, 2015, at 2:00 PM, Yuan Zhu <notifications@github.com> wrote:
> 
> In Parkinson/TasksAndSteps/APHParkinsonActivityViewController.m <https://github.com/ResearchKit/mPower/pull/6#discussion_r33512747>:
> 
> > +                ORKChoiceQuestionResult  *result = (ORKChoiceQuestionResult *)object;
> > +                    //
> > +                    //  stash the object away
> > +                    //
> > +                if (result.choiceAnswers.lastObject != nil) {
> > +                    NSDictionary  *dictionary = @{
> > +                                                  kMomentInDayChoiceAnswerKey : result.choiceAnswers.lastObject,
> > +                                                  kMomentInDayQuestionTypeKey : @(result.questionType),
> > +                                                  kMomentInDayIdentifierKey   : result.identifier,
> > +                                                  kMomentInDayStartDateKey    : result.startDate,
> > +                                                  kMomentInDayEndDateKey      : result.endDate,
> > +                                                  };
> > +                    NSUserDefaults  *defaults = [NSUserDefaults standardUserDefaults];
> > +                    [defaults setObject:dictionary forKey:kMomentInDayUserDefaultsKey];
> > +                    [defaults synchronize];
> > +                }
> Why do the stash? Is a feature which is missing from ResearchKit?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/ResearchKit/mPower/pull/6/files#r33512747>.
> 

The stash is done as a special case introduced at a late stage in the four Parkinson activities.

The idea is to inject a step at the beginning of the activity to ask the patient if they have taken their
medication prior to doing the  activity.   If they then do another activity within a time limit (currently
twenty minutes), they won’t get pestered again.

The original flow was that the results of that question would be uploaded with the other results.

Then the researchers asked us to upload the results even when the question didn’t get asked, so
we had to stash away the most recent results to be uploaded.

We could debate on whether ResearchKit should have that capability folded in.  The requirement
seems pretty esoteric.    Course, if other researchers ask for the same capability, that could be
the impetus to create the capability in a more formal way.

    Best Wishes,
        . . . . . . . .    Henry